Merge standard products feature into develop @W-18912410

sf-emmyzhang opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

This PR is merging all the standard product feature work into develop. See "How to Test-Drive This PR" below for all the relevant PRs

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

MRT Env: https://cc-sharks-standard-product.mobify-storefront-staging.com
PRs included in this feature branch:
- [Support standard product on PDP @W-12360908](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2518)
- [@W-12360935 Modify the Product Card to support standard products](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2581)
- [@W-12360935 2nd PR- fixing Edit button appearing briefly for standard products](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2649)
- [Support standard product as a child item in bundles @W-18790730](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2574)
- [Support standard product as a child item in sets @W-18790730](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2636)
- [@W-17118204 Verify einstein activity for Standard Product](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2650)
- [@W-18767088 refactor add to cart](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2664)

Test steps: 
## Happy path
2. Go to [home page](https://cc-sharks-standard-product.mobify-storefront-staging.com), clear site data, and refresh and press accept on consent tracking. Open Network tab and put in "cquotient". Then go to [standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/new-standard-product) and see that: viewProduct, pdp-similar-items, complete-the-set, viewReco requests are being made. viewReco will come up if you scroll down to the bottom of PDP. 
3. Have network tab still open. Add [standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/new-standard-product) to cart. See that addToCart cquotient call is made. And check out
4. Add [standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/new-standard-product) to wish list from the PDP and [PLP](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/search?q=standard%20collar%20shirt)
5. Add [bundle with standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/bundle-with-standard-product) to cart and checkout also add it to wish list
6. Add [set with standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/set-with-standard-product) to cart and checkout also add it to wish list
 
## Out of stock handling
1. See that [out of stock standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/just-a-standard-product) shows that it's out of stock
2. See that a [bundle with out of stock standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/Bundle%20with%20standard%20product%202) shows out of stock
3. See that a [set with out of stock standard product](https://cc-sharks-standard-product.mobify-storefront-staging.com/us/en-US/product/a-set-within-a-set) shows out of stock

Note: Adding an out of stock product in the wishlist, not limited to standard product, in the `develop` code currently gives a "Something went wrong" error message. This behavior existed before and is not blocking for this PR. 

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

shethj committed:
Handle SDK Client not initialized

shethj committed:
Added tests for useResolvedClient

shethj committed:
Handle missing SDK Client ShopperContext

shethj committed:
ShopperCustomers - Handle SDK Client no init

shethj committed:
ShopperExperience - Handle missing SDK Client

shethj committed:
ShopperGiftCertificates - Handle missing SDK Client

shethj committed:
ShopperLogin: Handle missing SDK client

shethj committed:
ShopperOrders - Handle Missing SDK Client

shethj committed:
ShopperProducts: Handle missing SDK Client

shethj committed:
ShopperPromotions: Handle missing SDK client

shethj committed:
ShopperSearch - Handle missing SDK Client

shethj committed:
ShopperSEO: Handle missing SDK Client

shethj committed:
ShopperStores: Handle missing SDK Client

shethj committed:
Fix linting errors

shethj committed:
Update useResolvedClient tests

shethj committed:
Update changelog

shethj committed:
Update changelog

shethj committed:
Extract CLIENT_KEY literal to constnt

shethj committed:
Remove additional hook for client validation

shethj committed:
Fix linting

shethj committed:
Add comments

shethj committed:
rename apiClients variable to add clarity

Co-authored-by: Kevin He <kevin.he@salesforce.com>
Signed-off-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>

shethj committed:
update variable name for apiClients

bendvc committed:
Initial Commit

Follow `plugin_einstein_api` implementation for event data creation.

bendvc committed:
Update CHANGELOG.md

bendvc committed:
Add some additional tests

sf-emmyzhang committed:
Support standard product on PDP @W-12360908 (#2518)

* update handleAddToCart and validate orderability

* remove bundle changes, template test, and project config

* add fallback if no image

* linting

* add validateOrderability tests

* reorganize tests

* fix lint errors

* update validateOrderability logic

* add product-detail tests

* fix modal image link

* add modal tests

* changelog

* pr suggestions

sf-emmyzhang committed:
Support standard product on PDP @W-12360908 (#2518)

* update handleAddToCart and validate orderability

* remove bundle changes, template test, and project config

* add fallback if no image

* linting

* add validateOrderability tests

* reorganize tests

* fix lint errors

* update validateOrderability logic

* add product-detail tests

* fix modal image link

* add modal tests

* changelog

* pr suggestions

bendvc committed:
PR Feedback

bendvc committed:
Fix bar reference to master.variantId

szirpesf committed:
feat: cursor rules for unit tests

bendvc committed:
Update change log

shethj committed:
Update sample query implementation in query.ts

sf-emmyzhang committed:
Resolve merge conflict

sf-emmyzhang committed:
Update files with conflicts

sf-emmyzhang committed:
omg so we're not adding these

sf-emmyzhang committed:
merge develop into standard product feature branch  (#2626)

* Handle SDK Client not initialized

* Added tests for useResolvedClient

* Handle missing SDK Client ShopperContext

* ShopperCustomers - Handle SDK Client no init

* ShopperExperience - Handle missing SDK Client

* ShopperGiftCertificates - Handle missing SDK Client

* ShopperLogin: Handle missing SDK client

* ShopperOrders - Handle Missing SDK Client

* ShopperProducts: Handle missing SDK Client

* ShopperPromotions: Handle missing SDK client

* ShopperSearch - Handle missing SDK Client

* ShopperSEO: Handle missing SDK Client

* ShopperStores: Handle missing SDK Client

* Fix linting errors

* Update useResolvedClient tests

* Update changelog

* implemented the voiceover feature for the email confirmation modal on PWA kit.

* Update changelog

* Extract CLIENT_KEY literal to constnt

* updated the chanege log file to point to current changes and saved changes to email confirmation modal code

* linted the files and built translations

* changed role and aria label tags to solve unsuccessful check issues

* Implemented changes to create more successful checks after facing a testing Library ElementError: Found multiple elements with the text: /check your email/i

* Linted my files

* Remove additional hook for client validation

* Fix linting

* Add comments

* rename apiClients variable to add clarity

Co-authored-by: Kevin He <kevin.he@salesforce.com>
Signed-off-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>

* update variable name for apiClients

* linted more files to fix unsuccessful check error

* linted more files to fix unsuccessful check error

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* taking off  the visible focus outline around the modal border

* Initial Commit

Follow `plugin_einstein_api` implementation for event data creation.

* Update CHANGELOG.md

* Update CHANGELOG.md

* Add some additional tests

* PR Feedback

* adding an a11y tag to changes made docuemented in the change log reading document

* Fix bar reference to master.variantId

* removed redundant aria label that repeating email confirmation title

* mitigating the need for additional translations

* feat: cursor rules for unit tests

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* Update change log

* Fixing bad merge

* Update sample query implementation in query.ts

* [Fix E2E Tests] Improve E2E Test and Tracking Consent Banner Handling (@W-18764173@) (#2575)

* Improve e2e tests and Tracking Consent banner handling

* Clean up

* More clean up

* Remove spacing changes

* PR Feedback

* Remove comments

* rules only

* Update translations (#2616)

* update translations

* @W-18541294@ Private client proxy updates (#2608)

* Ensure only requests to /shopper/auth/ are allowed by the SLAS private client proxy

* Remove console.logs

* Stop swallowing errors from SLAS

* Update CHANGELOG.md

* Fix login and logout

* Remove console logs

* Lint

* Add and fix tests

* Cleanup regex

---------

Signed-off-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>
Signed-off-by: Ben Chypak <bchypak@mobify.com>
Co-authored-by: Jainam Sheth <j.sheth@salesforce.com>
Co-authored-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>
Co-authored-by: Larnelle Ankunda <lankunda@salesforce.com>
Co-authored-by: Kevin He <kevin.he@salesforce.com>
Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Co-authored-by: Ben Chypak <bchypak@salesforce.com>
Co-authored-by: Ben Chypak <bchypak@mobify.com>
Co-authored-by: Shailesh Zirpe <szirpe+sf@salesforce.com>
Co-authored-by: Shailesh Zirpe <42553862+szirpesf@users.noreply.github.com>
Co-authored-by: Adam Raya <adamraya@users.noreply.github.com>
Co-authored-by: Alex Vuong <alex.vuong@salesforce.com>
Co-authored-by: vcua-mobify <47404250+vcua-mobify@users.noreply.github.com>

sf-shikhar-prasoon committed:
@W-12360935 Modify the Product Card to support standard products (#2581)

* remove 'edit'for std prd | Add unit tests

* update changelog

* lint fixes

* show Edit button on bundleProducts

* handle empty product bundles

* better determination of if-standardProduct

---------

Signed-off-by: sf-shikhar-prasoon <214730309+sf-shikhar-prasoon@users.noreply.github.com>

sf-emmyzhang committed:
udpate changelog to resolve conflict

sf-emmyzhang committed:
udpate changelog to resolve conflict 2

sf-emmyzhang committed:
udpate changelog to resolve conflict 3

sf-emmyzhang committed:
oops

sf-emmyzhang committed:
Merge develop into standard-product-support (#2633)

* Handle SDK Client not initialized

* Added tests for useResolvedClient

* Handle missing SDK Client ShopperContext

* ShopperCustomers - Handle SDK Client no init

* ShopperExperience - Handle missing SDK Client

* ShopperGiftCertificates - Handle missing SDK Client

* ShopperLogin: Handle missing SDK client

* ShopperOrders - Handle Missing SDK Client

* ShopperProducts: Handle missing SDK Client

* ShopperPromotions: Handle missing SDK client

* ShopperSearch - Handle missing SDK Client

* ShopperSEO: Handle missing SDK Client

* ShopperStores: Handle missing SDK Client

* Fix linting errors

* Update useResolvedClient tests

* Update changelog

* implemented the voiceover feature for the email confirmation modal on PWA kit.

* Update changelog

* Extract CLIENT_KEY literal to constnt

* updated the chanege log file to point to current changes and saved changes to email confirmation modal code

* linted the files and built translations

* changed role and aria label tags to solve unsuccessful check issues

* Implemented changes to create more successful checks after facing a testing Library ElementError: Found multiple elements with the text: /check your email/i

* Linted my files

* Remove additional hook for client validation

* Fix linting

* Add comments

* rename apiClients variable to add clarity

Co-authored-by: Kevin He <kevin.he@salesforce.com>
Signed-off-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>

* update variable name for apiClients

* linted more files to fix unsuccessful check error

* linted more files to fix unsuccessful check error

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* taking off  the visible focus outline around the modal border

* Initial Commit

Follow `plugin_einstein_api` implementation for event data creation.

* Update CHANGELOG.md

* Update CHANGELOG.md

* Add some additional tests

* PR Feedback

* adding an a11y tag to changes made docuemented in the change log reading document

* Fix bar reference to master.variantId

* removed redundant aria label that repeating email confirmation title

* mitigating the need for additional translations

* feat: cursor rules for unit tests

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* Update packages/template-retail-react-app/CHANGELOG.md

Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>

* Update change log

* Fixing bad merge

* Update sample query implementation in query.ts

* [Fix E2E Tests] Improve E2E Test and Tracking Consent Banner Handling (@W-18764173@) (#2575)

* Improve e2e tests and Tracking Consent banner handling

* Clean up

* More clean up

* Remove spacing changes

* PR Feedback

* Remove comments

* rules only

* Update translations (#2616)

* update translations

* @W-18541294@ Private client proxy updates (#2608)

* Ensure only requests to /shopper/auth/ are allowed by the SLAS private client proxy

* Remove console.logs

* Stop swallowing errors from SLAS

* Update CHANGELOG.md

* Fix login and logout

* Remove console logs

* Lint

* Add and fix tests

* Cleanup regex

* skip changelog

---------

Signed-off-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>
Signed-off-by: Larnelle Ankunda <lankunda@salesforce.com>
Signed-off-by: Ben Chypak <bchypak@mobify.com>
Co-authored-by: Jainam Sheth <j.sheth@salesforce.com>
Co-authored-by: Jainam Sheth <99490559+shethj@users.noreply.github.com>
Co-authored-by: Larnelle Ankunda <lankunda@salesforce.com>
Co-authored-by: Kevin He <kevin.he@salesforce.com>
Co-authored-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Co-authored-by: Ben Chypak <bchypak@salesforce.com>
Co-authored-by: Ben Chypak <bchypak@mobify.com>
Co-authored-by: Shailesh Zirpe <szirpe+sf@salesforce.com>
Co-authored-by: Shailesh Zirpe <42553862+szirpesf@users.noreply.github.com>
Co-authored-by: Adam Raya <adamraya@users.noreply.github.com>
Co-authored-by: Alex Vuong <alex.vuong@salesforce.com>
Co-authored-by: vcua-mobify <47404250+vcua-mobify@users.noreply.github.com>

sf-emmyzhang committed:
Support standard product as a child item in bundles @W-18790730 (#2574)

* init commit

* skip variation check for standard product

* update modal logic

* add product view test

* add to cart modal test

* add product detail test

* update add to cart modal test

* add product util test

* update changelog

* fix edit bundle

* lint

jeremy-jung1 committed:
Merge branch 'develop' into merge-develop-into-standard-product-support-1

jeremy-jung1 committed:
Handle standard product shape in sendAddToCart

jeremy-jung1 committed:
Update CHANGELOG.md

sf-emmyzhang committed:
Support standard product as a child item in sets @W-18790730 (#2636)

* fix adding set to cart

* updae changelog

* udpate util test

jeremy-jung1 committed:
Add test

jeremy-jung1 committed:
Lint tests

sf-kyle-wright committed:
@W-18767088 refactor add to cart

sf-kyle-wright committed:
@W-18767088 refactor add to cart

sf-kyle-wright committed:
@W-18767088 fix lint errors

jeremy-jung1 committed:
Merge branch 'standard-product-support' into W-17118204-verify-einstein-activity

jeremy-jung1 committed:
Merge branch 'develop' into W-17118204-verify-einstein-activity

jeremy-jung1 committed:
Merge pull request #2650 from SalesforceCommerceCloud/W-17118204-verify-einstein-activity

@W-17118204 Verify einstein activity for Standard Product

sf-kyle-wright committed:
@W-18767088 do not move handleChildProductValidation

sf-kyle-wright committed:
Merge branch 'standard-product-support' into t/cc-shark/W-18767088/refactor-add-to-cart

Signed-off-by: Kyle Wright <30636085+sf-kyle-wright@users.noreply.github.com>

sf-shikhar-prasoon committed:
@W-12360935 2nd PR- fixing Edit button appearing briefly for standard products (#2649)

* fix Edit button appearing briefly
* fix tests

sf-kyle-wright committed:
@W-18767088 utils rename

sf-kyle-wright committed:
Merge pull request #2664 from SalesforceCommerceCloud/t/cc-shark/W-18767088/refactor-add-to-cart

@W-18767088 refactor add to cart

cc-prodsec commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/a1686db9-aa9b-44f0-a0ac-64231ee32640?refs=c591eaa9ca21b54a5a2f0fe31adbde68b650c583&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/a1686db9-aa9b-44f0-a0ac-64231ee32640/license?refs=c591eaa9ca21b54a5a2f0fe31adbde68b650c583&source=prChecksComment)

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

Signed-off-by: sf-emmyzhang <emmyzhang@salesforce.com>

sf-emmyzhang committed:
Revert CHANGELOG.md

Signed-off-by: sf-emmyzhang <emmyzhang@salesforce.com>

sf-emmyzhang committed:
increase bundle size for ci error

sf-shikhar-prasoon commented on packages/template-retail-react-app/CHANGELOG.md:
```diff
@@ -5,7 +5,13 @@
+- Support standard product on PDP [#2518](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2518)
+- Removed the "Edit" button from product cards of Standard Products [#2581](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2581)
+- Support standard product as a child item in bundles [#2574](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2574)
+- Verify einstein activity with standard product [#2650](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2650)
```
Should we be adding this line to changelog?
I don't think we're adding any code changes related to einstein activity?

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

Signed-off-by: sf-emmyzhang <emmyzhang@salesforce.com>

kzheng-sfdc commented on packages/template-retail-react-app/app/components/product-view-modal/bundle.test.js:
```diff
@@ -166,6 +166,20 @@ test('bundle product view modal disables update button when quantity exceeds chi
+        expect(screen.getAllByText('Sleeveless Pleated Floral Front Blouse')).toHaveLength(2)
+        expect(screen.getAllByText('Swing Tank')).toHaveLength(2)
+        expect(screen.getAllByText('Pull On Neutral Pant')).toHaveLength(2)
+    })
```
is any of these products a standard/simple product? 
It's not clear to me whether we have new test case to exercise the inclusion of a simple product in a bundle.

sf-emmyzhang replied:
No, they are not simple products. These were added to make the test more stable.

kzheng-sfdc replied:
OK, then can we add a new test case to cover simple products?

kzheng-sfdc commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -169,8 +182,8 @@ const ProductView = forwardRef(
-        const isProductASet = product?.type.set
-        const isProductABundle = product?.type.bundle
+        const isProductASet = product?.type?.set
+        const isProductABundle = product?.type?.bundle
```
When would product.type be null? A standard product?

sf-emmyzhang replied:
For a standard product `product.type.item` should be true. `product.type` could be null or undefined when the API doesn't return type information. This is to safely handle inconsistent product data so it doesn't crash with a "Cannot read property 'bundle' of null" error.

kzheng-sfdc commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -125,8 +125,21 @@ const ProductView = forwardRef(
+                // If product has variations, a variant must be selected
+                if (product?.variationAttributes?.length > 0 && !variant) {
+                    return false
+                }
```
If product is a variant, would not not be guarded by variant?.orderable check?

sf-emmyzhang replied:
Good question, so this check here is slightly different from the variant.orderable check below in that the former verifies if a variant has been selected at all (if any variant exists), not checking if the variant is orderable. The latter validates the orderability of the selected variant (or standard product). 

The first check is primarily a guard to prevent attempting to validate orderability when no variant has been selected in the first place. So in the scenario where there is a product with variations but no variations selected, this check will fail and prevent further unnecessary checks.

You could say that if a variant is selected but not orderable, the variant?.orderable check would catch it. I think the separation makes the logic clearer and provides betterr error handling for different failure scenarios (no selection vs. invalid selection). Feel free to chime in with differing implementation options

kzheng-sfdc replied:
Got it. Thanks for the explanation - since this is an improvement over existing variant logic, that's why I was wondering.

kzheng-sfdc commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -292,7 +305,7 @@ const ProductView = forwardRef(
                 }
                 try {
-                    const itemsAdded = await addToCart([{variant, quantity}])
+                    const itemsAdded = await addToCart([{variant: variant || product, quantity}])
```
does addToCart([{variant, product, quantity}]) not work? I saw that handleAddToCart in product-view handles item.product.

sf-emmyzhang replied:
Good catch, the auto merge combined the two and changed the original implementation.

kzheng-sfdc commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -405,6 +418,16 @@ const ProductView = forwardRef(
+            ) {
+                onVariantSelected(product, null, childOfBundleQuantity || quantity)
+            }
+        }, [product, childOfBundleQuantity, quantity, isProductPartOfBundle, isProductPartOfSet])
```
Is this effect and the one above mutually exclusive? Can we combine them?

sf-emmyzhang replied:
I think these need to remain separate as the first one handles variants and this one specifically handled bundles/set items. Keeping them separate might also make it easier to read.

kzheng-sfdc replied:
Looking at the code again, it seems that we want to invoke the 2nd effect only when it's a simple product, but why are we calling onVariantSelected on a simple product? Doesn't it by design _not_ to have variantions to select from?

sf-emmyzhang replied:
For just a standard product, this effect should not run. For a standard product within a bundle/set, yes this effect is run. This effect is saying if this is a standard child product within a bundle/set, automatically select it with the specified quantity. And bc a standard child product has nothing for the user to select, the code automatically calls onVariantSelected(product, null, childOfBundleQuantity || quantity) (note that null is passed in for the variant value) to include them in the bundle selection.

This helps to ensure that standard products in bundles are automatically included without requiring user interaction, while variant products still require the user to make selections.

kzheng-sfdc commented on packages/template-retail-react-app/app/hooks/use-add-to-cart-modal.js:
```diff
@@ -133,7 +133,10 @@ export const AddToCartModal = () => {
+                                                <img
+                                                    src={bundleImage?.link || ''}
+                                                    alt={bundleImage?.alt || ''}
+                                                />
```
Would it make more sense to not render the image if it's not bundle?

sf-emmyzhang replied:
The current code seem to already conditionally render based on isProductABundle. The current state is that images are not rendered if not provided, but yeah I agree that it's not great to have an empty img tag. I think the better approach is to add a placeholder image. I'll add a todo comment here to address that.

sf-emmyzhang committed:
update use add to cart modal unit test

kzheng-sfdc commented on packages/template-retail-react-app/app/hooks/use-add-to-cart-modal.js:
```diff
@@ -239,7 +246,10 @@ export const AddToCartModal = () => {
+                                                        <img
+                                                            src={image?.link || ''}
+                                                            alt={image?.alt || ''}
+                                                        />
```
same comment as above

kzheng-sfdc commented on packages/template-retail-react-app/app/pages/product-detail/index.jsx:
```diff
@@ -161,7 +166,11 @@ const ProductDetail = () => {
     let bundleChildVariantIds = ''
     if (isProductABundle)
         bundleChildVariantIds = Object.keys(childProductSelection)
```
Can we rename this to bundleChildProductIds, as it's no longer just variant?

kzheng-sfdc commented on packages/template-retail-react-app/app/pages/product-detail/index.jsx:
```diff
@@ -183,7 +192,7 @@ const ProductDetail = () => {
         product.bundledProducts.forEach(({product: childProduct}, index) => {
             const matchingChildProduct = bundleChildrenData.data.find(
-                (bundleChild) => bundleChild.master.masterId === childProduct.id
+                (bundleChild) => bundleChild?.master?.masterId === childProduct.id
```
Why is this change necessary? Does standard product not have masterId, which I assume is the parent product id?

sf-emmyzhang replied:
No standard product does not have a master property. Since we need to support standard products in bundles, the bundle child structure would be different. The optional chaining is needed to prevent type errors. Also I'm referenced this [doc](https://salesforce.quip.com/KLmCAmbGFz9a) for the shape of standard products if you're interested.

kzheng-sfdc replied:
So in another word, this loop does not apply to simple products? It's so because the inventory of a simple product doesn't change the same way variants do?

sf-emmyzhang replied:
Correct me if I'm wrong, but I'm assuming your question is hinting that since this filtering logic is designed for variant products and not standard products we should add another check designed specifically for standard (bundleChild?.productId === childProduct.id)?

kzheng-sfdc replied:
Looking at next lines, matchingChildProduct is then used to provide the inventory number. So for standard product, we won't find any matchingChildProduct, will that be OK to skip the inventory number setting?

sf-emmyzhang replied:
Hmm, good point. I don't know if it's ok to skip inventory number check for standard products, I'm leaning towards no and that we should be checking as not checking for standard would mean that standard child product for bundles may not have accurate inventory information. @jeremy-jung1 Can you help confirm?

jeremy-jung1 replied:
Hm I don't see a use of the `inventory` field here being used in the codebase. The more useful inventory check is actually done in `useDerivedProduct`, which allows for inventoryMessage like "Out of stock" to show up. I'll make a change so that this message toggle is being applied to standard product.

jeremy-jung1 replied:
Added the support for standard product in these commits ([0a45ee8](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2697/commits/0a45ee86950407529eabda06b9eae245b52e8354), [5707bb5](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2697/commits/5707bb5346c2b3771911b5d3bc6096f10d007bc4)), [c915f7c](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2697/commits/c915f7ccbe7ec0eff161f99708576043d64abb51)

Before with an out of stock standard product: 
![Screenshot 2025-07-05 at 4 44 47 PM](https://github.com/user-attachments/assets/a5cccd93-4cd9-41c8-981c-5829c7a771e4)

After: 
![Screenshot 2025-07-05 at 4 37 53 PM](https://github.com/user-attachments/assets/91009331-54b3-4c23-a284-52e7d74dc81f)
![Screenshot 2025-07-05 at 4 38 10 PM](https://github.com/user-attachments/assets/191ded6f-1578-4451-9187-06534cfb369e)

kzheng-sfdc commented on packages/template-retail-react-app/app/utils/add-to-cart-utils.js:
```diff
@@ -0,0 +1,154 @@
+ * @param {Function} showError
+ * @returns {Promise<Array>|undefined}
+ */
+export const handleAddToCart = async (
```
Is this related to standard product, or result of merging?

sf-emmyzhang replied:
No, this was part of the add to cart util refactoring work. Please see https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2664, or view How to Test-Drive This PR section of the PR description.

sf-emmyzhang committed:
revert auto merge change

sf-emmyzhang committed:
add image placeholder todo

sf-emmyzhang committed:
rename bundleChildVariant

sf-emmyzhang committed:
revert refactor work

sf-emmyzhang committed:
add bundle test

sf-emmyzhang committed:
lint

sf-emmyzhang committed:
update test

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

sf-emmyzhang committed:
lint

jeremy-jung1 committed:
Handle out of stock for standard product

jeremy-jung1 committed:
tests and linting

jeremy-jung1 committed:
handle lack of type

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

jeremy-jung1 committed:
Fix child standard product item in set 400 on addToCart

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

Signed-off-by: sf-emmyzhang <emmyzhang@salesforce.com>

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

patricksullivansf commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -292,7 +305,7 @@ const ProductView = forwardRef(
                 }
                 try {
-                    const itemsAdded = await addToCart(variant, quantity)
+                    const itemsAdded = await addToCart([{product, variant, quantity}])
```
hi, @sf-emmyzhang we just removed this change in #2729. Is there a benefit to adding it back?

sf-emmyzhang replied:
Product is part of standard product changes and is needed for standard products to work

patricksullivansf replied:
Hmm... before BOPIS or standard-product the `product` was added as a lambda in the parent container `addToCart` handler in `pages/product-detail`.  And a `childProduct` was added in the parent for product sets and bundles.  My worry is with this change back that bundles and sets will not work properly.

jeremy-jung1 replied:
Updated the code to account for both features ([commit](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2697/commits/fd8b3bd02f23d82187191c5d7746e61406b64106))

patricksullivansf replied:
nit: for the record I still don't think this change is needed nor advisable. We're changing the data contract between `product-view` and `product-detail`, which implies more merge work for custom implementations. There should be a good reason.

jeremy-jung1 replied:
Yeah It's a little difficult to account for multiple product types and the code is starting feel a little bit spaghetti. Some refactoring would be really nice, but that's a larger undertaking. This was actually [SPIKED](https://salesforce.quip.com/T9YQAo4m6ftz) a bit but the more features are added week after week the more we stray away from it.

But what data contract do you mean specifically? If you mean the change of the arguments from (variant, quantity) to ([{product, variant, quantity}]) I think it's fine because before #2729 it still used to be of the shape ([{variant ,quantity}]). The good thing about passing around objects instead of separate arguments is that additions are usually backwards compatible. And the release hasn't been cut yet so there wasn't much of a change across these PRs.

patricksullivansf commented on packages/template-retail-react-app/app/pages/product-detail/index.jsx:
```diff
@@ -696,14 +706,10 @@ const ProductDetail = () => {
-                                                                      quantity
-                                                                  }
-                                                              ])
+                                                        ? (productSelectionValues) =>
+                                                              handleAddToCart(
+                                                                  productSelectionValues
+                                                              )
```
Guessing we don't want this change either as it would associate product sets with the parent instead of the child, no?

sf-emmyzhang replied:
@jeremy-jung1 Can you help explain this change? Is this related to the 400 error?

jeremy-jung1 replied:
Yes this was the fix for a 400 error that occurred when trying to add a standard product child item in a set to cart. The childProduct is passed in via the `<ProductView/>` component because it received the `childProduct` via the `product` prop and passes that in when calling addToCart. 

Before the parameters passed into handleAddToCart was just `(variant, quantity)` which isn't the correct shape as how the function is used in ProductView.

patricksullivansf commented on packages/template-retail-react-app/app/utils/add-to-cart-utils.js:
```diff
@@ -0,0 +1,154 @@
+ */
+
+/**
+ * Handles adding products to cart and sending data to Einstein.
```
I don't see BOPIS support here. do we want to add a comment that states exactly that?

vmarta commented:
There's a bug with adding to wishlist when you're initially a guest shopper. After logging in, we automatically tried to mutate/update the wishlist but ran into a js undefined error.

What the expected result is:
- successful login ← it's good
- and also successfully added to the wishlist ← failed here

Here's a video to reproduce the bug:
 
https://github.com/user-attachments/assets/e66632c0-aa4a-43fa-8200-eb39c106a07b

sf-emmyzhang committed:
pr suggestions

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

sf-emmyzhang commented:
> There's a bug with adding to wishlist when you're initially a guest shopper. After logging in, we automatically tried to mutate/update the wishlist but ran into a js undefined error.

I followed these steps to repro, on my local but it seems to be working for me. Did you test on the standard product MRT env? It's possible that we need to deploy again.

sf-emmyzhang committed:
lint and test failure

vmarta commented on packages/template-retail-react-app/app/pages/product-detail/index.jsx:
```diff
@@ -158,23 +158,27 @@ const ProductDetail = () => {
+        bundleChildProductIds = Object.keys(childProductSelection)
+            ?.map(
+                (key) =>
+                    childProductSelection[key].variant?.productId ||
+                    childProductSelection[key].product?.id
```
Non-blocking: I see a lot of places where we check for variant or standard products... it requires us to do this check in all places that need it.

In the future, it'd be great to normalize the product data in 1 place initially. For example, according to AI:

```js
// At the boundary - normalize immediately
const normalizeProductSelection = (variant, product) => {
  // Always return same shape, no matter what input look like
  return {
    id: variant?.productId || product?.id,
    price: variant?.price || product?.price,
    name: variant?.name || product?.name,
    orderable: variant?.orderable || product?.inventory?.orderable,
    stockLevel: variant?.stockLevel || product?.inventory?.stockLevel,
    isStandardProduct: !variant,
    isVariantProduct: !!variant,
    originalVariant: variant,
    originalProduct: product
  }
}

// Now everywhere else in code - simple!
const addToCart = (normalizedProduct, quantity) => {
  return {
    productId: normalizedProduct.id,  // No more conditionals!
    price: normalizedProduct.price,   // Simple!
    quantity
  }
}

// Validation becomes obvious
const validateOrderability = (normalizedProduct, quantity) => {
  return normalizedProduct.orderable && 
         quantity > 0 && 
         quantity <= normalizedProduct.stockLevel
}

```

vmarta commented:
> > There's a bug with adding to wishlist when you're initially a guest shopper. After logging in, we automatically tried to mutate/update the wishlist but ran into a js undefined error.
> 
> I followed these steps to repro, on my local but it seems to be working for me. Did you test on the standard product MRT env? It's possible that we need to deploy again.

@sf-emmyzhang yes, I tested on the MRT environment listed in the PR description.

jeremy-jung1 committed:
Fix standard product bundle add to cart for pick up in store?. And linting

sf-emmyzhang committed:
Merge branch 'develop' into standard-product-support

jeremy-jung1 committed:
Ensure toast has a string to show

Toast needs a string to show the message properly. If it just receives an Error Object, it will try to show that in DOM which causes a blank screen

jeremy-jung1 committed:
linting

vmarta reviewed (approved):
With the latest deployed env, I'm still seeing the issue with adding to wishlist as a guest user. Perhaps it's a race condition that I was able to run into consistently? 

It's strange how the Welcome message does not have my name in it. Maybe it's related to the js undefined error somehow. 

https://github.com/user-attachments/assets/62b4cdbf-647f-4290-9aeb-406f7d0b6396

But anyways, let's create a ticket (or however you'd like to track it).

Other than that, this PR looks good to me.

ddiazccrz commented on packages/template-retail-react-app/CHANGELOG.md:
```diff
@@ -7,7 +7,13 @@
 - Update latest translations for all languages [#2616](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2616)
 - Added support for Buy Online Pick up In Store (BOPIS) [#2646](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2646)
+- Support standard product on PDP [#2518](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2518)
+- Removed the "Edit" button from product cards of Standard Products [#2581](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2581)
```
Small knit, but maybe just put the last PR (this one) in the changelog, i dont think we ended up adding all the statements for the other features.

ddiazccrz commented on packages/template-retail-react-app/CHANGELOG.md:
```diff
@@ -7,7 +7,13 @@
+- Verify einstein activity with standard product [#2650](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2650)
+- Support standard product as a child item in sets [#2636](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2636)
 - Load active data scripts on demand only [#2623](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2623)
+- Refactor Add to Cart in PDP [#2664](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2664)
```
This PR did not make it in to develop, so this needs to be pulled
